dll/win32/shell32/shlfolder.cpp

```diff
@@ -266,59 +266,80 @@ HRESULT SHELL32_CompareDetails(IShellFolder2* isf, LPARAM lParam, LPCITEMIDLIST
     return MAKE_COMPARE_HRESULT(ret);
 }
 
-void AddClassKeyToArray(const WCHAR * szClass, HKEY* array, UINT* cKeys)
+LSTATUS AddClassKeyToArray(const WCHAR * szClass, HKEY* array, UINT* cKeys)
 {
     if (*cKeys >= 16)
-        return;
+        return ERROR_MORE_DATA;
 
     HKEY hkey;
     LSTATUS result = RegOpenKeyExW(HKEY_CLASSES_ROOT, szClass, 0, KEY_READ | KEY_QUERY_VALUE, &hkey);
-    if (result != ERROR_SUCCESS)
-        return;
-
-    array[*cKeys] = hkey;
-    *cKeys += 1;
+    if (result == ERROR_SUCCESS)
+    {
+        array[*cKeys] = hkey;
+        *cKeys += 1;
+    }
+    return result;
 }
 
-void AddFSClassKeysToArray(PCUITEMID_CHILD pidl, HKEY* array, UINT* cKeys)
+void AddFSClassKeysToArray(UINT cidl, PCUITEMID_CHILD_ARRAY apidl, HKEY* array, UINT* cKeys)
 {
+    //learn.microsoft.com/en-us/windows/win32/shell/fa-associationarray#about-association-arrays
+
+    ASSERT(cidl >= 1 && apidl);
+    PCUITEMID_CHILD pidl = apidl[0];
     if (_ILIsValue(pidl))
     {
+        WCHAR buf[MAX_PATH], *name;
         FileStructW* pFileData = _ILGetFileStructW(pidl);
-        LPWSTR extension = PathFindExtension(pFileData->wszName);
+        if (pFileData)
+        {
+            name = pFileData->wszName;
+        }
+        else
+        {
+            _ILSimpleGetTextW(pidl, name = buf, _countof(buf));
+        }
+        LPWSTR extension = PathFindExtension(name);
 
```

Comment:
As far as I can see in the current function, `extension` can be made pointer to const wchar since the pointed contents don't get modified.
```suggestion

        LPCWSTR extension = PathFindExtensionW(name);
```